pkg/meta/model/job_args.go

```diff
@@ -1295,3 +1308,376 @@ func (a *FlashbackClusterArgs) decodeV1(job *Job) error {
+		if err := job.DecodeArgs(&indexNames, &ifExists); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	a.IndexNames = indexNames
+	a.IfExists = ifExists
+	return nil
+}
+
+func (a *DropIndexArgs) getArgsV1(*Job) []any {
+	if len(a.IndexNames) == 1 {
+		return []any{a.IndexNames[0], a.IfExists[0]}
+	}
+	return []any{a.IndexNames, a.IfExists}
+}
+
+// GetDropIndexArgs gets the drop index args.
+// It's used for both drop index and rollback add index.
+func GetDropIndexArgs(job *Job) (*DropIndexArgs, error) {
+	return getOrDecodeArgs(&DropIndexArgs{}, job)
+}
+
+func (a *DropIndexArgs) getFinishedArgsV1(*Job) []any {
+	// This is to make the args compatible with old logic:
+	// 1. For drop index, arguments are [CIStr, bool, int64, []int64].
+	// 2. For rollback add index, arguments are [[]CIStr, []bool, []int64].
+	if a.IsRollback {
+		return []any{a.IndexNames, a.IfExists, a.IndexIDs}
+	}
+	return []any{a.IndexNames[0], a.IfExists[0], a.IndexIDs[0], a.PartitionIDs, a.IsVector}
+}
+
+// GetFinishedDropIndexArgs gets the drop index args.
+// It's used for both drop index and rollback add index.
+func GetFinishedDropIndexArgs(job *Job) (*DropIndexArgs, error) {
+	if job.Version == JobVersion1 {
+		indexNames := make([]pmodel.CIStr, 1)
+		ifExists := make([]bool, 1)
+		indexIDs := make([]int64, 1)
+		var partitionIDs []int64
+		isVector := false
+
+		// See getFinishedArgsV1 for why we may need to decode twice.
+		if err := job.DecodeArgs(&indexNames, &ifExists, &indexIDs, &isVector); err != nil {
+			if err := job.DecodeArgs(&indexNames[0], &ifExists[0], &indexIDs[0], &partitionIDs, &isVector); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+
+		return &DropIndexArgs{
+			IndexNames:   indexNames,
+			IfExists:     ifExists,
+			IndexIDs:     indexIDs,
+			PartitionIDs: partitionIDs,
+			IsVector:     isVector,
+		}, nil
+	}
+	return getOrDecodeArgsV2[*DropIndexArgs](job)
+}
+
+// IndexArg is the argument for single add index operation
+// For NonPK, used fields are (listed in order of v1)
+// Unique, IndexName, IndexPartSpecifications, IndexOption, SQLMode, Warning, Global
+// For PK
+// Unique, IndexName, IndexPartSpecifications, IndexOptions, HiddelCols, Global
+type IndexArg struct {
+	// Global is never used, we only use Global in IndexOption. Can be deprecated later.
+	Global                  bool                          `json:"-"`
+	Unique                  bool                          `json:"unique,omitempty"`
+	IndexName               pmodel.CIStr                  `json:"index_name,omitempty"`
+	IndexPartSpecifications []*ast.IndexPartSpecification `json:"index_part_specifications"`
+	IndexOption             *ast.IndexOption              `json:"index_option,omitempty"`
+	HiddenCols              []*ColumnInfo                 `json:"hidden_cols,omitempty"`
+
+	// For vector index
+	FuncExpr string `json:"func_expr,omitempty"`
+	IsVector bool   `json:"is_vector,omitempty"`
```

Comment:
For non-pointer types, `omitempty` doesn't work. Is this expected?

Reply:
They will be omitted if they are zero values.